Cached

pmrv opened the pull request.

pmrv committed:
WIP

freyso commented on h5io_browser/base.py:
```diff
@@ -229,6 +230,44 @@ def _merge_nested_dict(main_dict: dict, add_dict: dict) -> dict:
+@contextlib.contextmanager
+def WithOpenHDF(filename: str, mode: str = "r", swmr: bool = False):
+    filename = os.path.realpath(os.path.abspath(filename))
+    with _FILE_HANDLE_CACHE_LOCK:
+        file_is_open = filename in _FILE_HANDLE_CACHE
+    # File already cached, someone above us opened it and will clean it up, we just need to pass it down
+    if file_is_open:
+        # print("EVERYBODY HIT THE CACHE")
+        with _FILE_HANDLE_CACHE_LOCK:
+            handle = _FILE_HANDLE_CACHE[filename]
```
This is a classic mistake in multithreading. You can have a race condition between the check and the retrieval. You must do the `filename in _FILE_HANDLE_CACHE` and `handle =  _FILE_HANDLE_CACHE[key]` within the same locking.

pmrv replied:
Naturally. :|

freyso commented on h5io_browser/base.py:
```diff
@@ -229,6 +230,44 @@ def _merge_nested_dict(main_dict: dict, add_dict: dict) -> dict:
+from collections import defaultdict
+_FILE_HANDLE_CACHE_COUNTER = defaultdict(int)
+_FILE_HANDLE_CACHE = {}
+import threading
+_FILE_HANDLE_CACHE_LOCK = threading.RLock()
```
Maybe we should forget about threading for the time being. The h5 library by itself must be compiled in a special mode to be thread-save, and the conda-version usually isn't.

pmrv replied:
Ah, didn't know about that.  Anyway, the chances of pyiron becoming thread safe in the near future are comfortably close to zero

freyso commented on h5io_browser/base.py:
```diff
@@ -229,6 +230,44 @@ def _merge_nested_dict(main_dict: dict, add_dict: dict) -> dict:
+        # print("EVERYBODY HIT THE CACHE")
+        with _FILE_HANDLE_CACHE_LOCK:
+            handle = _FILE_HANDLE_CACHE[filename]
+        assert mode == handle.mode
```
I have an easy solution for this - let's talk on Monday.

freyso commented on h5io_browser/base.py:
```diff
@@ -229,6 +230,44 @@ def _merge_nested_dict(main_dict: dict, add_dict: dict) -> dict:
+        with _FILE_HANDLE_CACHE_LOCK:
+            handle = _FILE_HANDLE_CACHE[filename]
+        assert mode == handle.mode
+        if swmr != handle.swmr_mode:
```
swmr can be enabled on an open file. It doesn't hurt to have it enabled, so if either the old or new

freyso commented on h5io_browser/base.py:
```diff
@@ -229,6 +230,44 @@ def _merge_nested_dict(main_dict: dict, add_dict: dict) -> dict:
+            with _FILE_HANDLE_CACHE_LOCK:
+                _FILE_HANDLE_CACHE[filename] = handle
+            try:
+                # Monkey patch close so that downstream code can use the handle as a context manager without closing the
+                # file we want to cache
+                handle.close = lambda: None
```
This will not help.

- If the competing close is from an outer context, there is no need, because the outer context will close the file after the inner context has ended. 
- Closing the file manually within the context is against intended use.
- Protecting against close from concurrent threads works better via reference counting. Manipulation of handle is not thread-safe, so race-conditions are garantueed.

pmrv replied:
> * If the competing close is from an outer context, there is no need, because the outer context will close the file after the inner context has ended. 
> * Closing the file manually within the context is against intended use.

The issue is code that is unaware that the context is used (somewhere up the call stack).  Say I have two functions `foo` and `bar`.  The use the `_open_hdf` function to obtain a handle and use it as a context manager, because they want to make sure the file is closed.
```python
def foo(...):
  with _open_hdf() as handle:
    ...
def bar(...):
  with _open_hdf() as handle:
    ...
```
(I'm leaving out the details but assume they open the same file).
If I call them within the new context 
```python
with WithOpenHDF():
  foo()
  bar()
```
both calls receive the same handle from `_FILE_HANDLE_CACHE`, but the `with ... as handle:` in `foo()` will close the handle after the block. Our context manager doesn't get run though, so we cannot update the cache (and if we did it would subvert the whole point of the cache), which means `bar()` will get a cached but closed file handle.

> Protecting against close from concurrent threads works better via reference counting. Manipulation of handle is not thread-safe, so race-conditions are garantueed.

Agree, but the only way without monkey patching `close` I see is making a wrapping proxy object that behaves just like a hdf handle, except it has a reference counting `close`.  Would be doable, but annoying.  Maybe giving up on thread safety is the saner option then.

freyso reviewed (commented):
Some first few thoughts. I had overlooked your reply this morning, so I have a competing effort. We should sit together on Monday in pyiron meeting to discuss both.

pmrv committed:
Fix race conditions

Do the check and retrieve in the same lock context.  Monkey patching as well.

pmrv committed:
Remove threading locks because neither HDF5 nor pyiron is threadsafe

Also rename context manager

pmrv committed:
Add new function open_hdf and only use cache there

Wrap handles in a nullcontext so that calling code can use

>>> with open_hdf(...) as handle:
...   ...

as before without accidentally closing the file handle

pmrv committed:
Fix a missing close bug in _read_hdf

If passed a str, the previous code opened the file, but didn't close or
return it, therefor leaking it.

pmrv committed:
Do not monkey patch

Instead I'll add a code path in open_hdf that checks if the file is
closed and reopens under a warning as necessary.

pmrv committed:
Wrap all tests inside CachedHDF

pmrv committed:
Pull accessing the cache into its own function

Treat compatible modes correctly and do not mix a/r+.

pmrv committed:
Use open_hdf everywhere

pmrv committed:
Remove debug counter

pmrv committed:
Workaround 3.11 not having enterContext

pmrv committed:
Do not use enterContext

Using this method the clean up is called after tearDown and that causes
an error on windows.
